### PR TITLE
ARROW-4081: [Go] Sum methods panic when the array is empty

### PR DESCRIPTION
**Pull request by alexandreyc**
Nothing fundamental but more idiomatic and coherent.

### CODE REVIEW

**Review by sbinet — APPROVED** [review]

**Comment by wesm** [comment]
@alexandreyc it looks like you're developing out of your master branch and so some unrelated changes got exposed here

**Comment by wesm** [comment]
maybe make this PR the fix for ARROW-4081? It's fine to combine those stylistic changes with this I think

**Comment by alexandreyc** [comment]
Yes ok, we will use this PR for ARROW-4081. I just need a bit more work to be sure for the fix. I will do it pretty soon.

**Comment by alexandreyc** [comment]
The issue is not specific to macOS since I've been able to reproduce it on Linux.

When reading the code it's clear why we panic when an empty array is passed: we take a pointer on the first element of the array to pass to low-level optimized function.

See for instance: https://github.com/apache/arrow/blob/d90c1597ce488d27aca88bcc33bfffe1e96a1ac6/go/arrow/math/float64_avx2_amd64.go#L35

What do you think about this fix? 

@sbinet @stuartcarnie

File: go/arrow/math/type_test.go.tmpl [hunk 1]
```diff
@@ -33,6 +33,14 @@ func Test{{.Name}}Funcs_Sum(t *testing.T) {
 	assert.Equal(t, res, {{.Type}}(49995000))
 }
 
+func Test{{.Name}}Funcs_SumEmpty(t *testing.T) {
+	b := array.New{{.Name}}Builder(memory.NewGoAllocator())
+	vec := b.New{{.Name}}Array()
+	b.Release()
```
Comment (sbinet) [inline go/arrow/math/type_test.go.tmpl]: I'd put it at line 38 as a `defer b.Release()` instead.
also: `vec` is leaked. we should probably have a `defer vec.Release()` somewhere as well.

perhaps we should use a CheckedAllocator:

```go
mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
defer mem.AssertSize(t, 0)
```

**Review by sbinet — COMMENTED** [review]
the general approach SGTM.

**Comment by alexandreyc** [comment]
Should be good now.

**Review by sbinet — APPROVED** [review]

**Pull request closed without merging**